File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Motherboard.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Objects;

import static android.os.Build.VERSION_CODES.Q;

public class Motherboard extends Part implements Parcelable {
	private boolean ecc; // has error correction coding (all chips should tbh)
	private String chipset;
	private String formFactor;
	private ArrayList<String> m2slots; // no clue what this is
	private int maxMemSupport;
	private int memSlots;
	private ArrayList<String> compatibleMem;
	private String memType; // pretty much always DDR4
	private int mSATA_slotCount; // love SSDs
	private ArrayList<String> incEthernetSupp;
	private String incVideo; // no idea what this is
	private ArrayList<CountedString> PCISlotList; // had to create a subobject for this
	private boolean hasRAID; // has "Redundant Array of Inexpensive Disks"
	private int sata6gbpsCount; // no clue
	private int gen2USBCount;
	private ArrayList<CountedString> gen32USBcount; // I don't feel like making a new USB object to only be used once
	private boolean isWireless;

	public Motherboard(String model, String manufacturer) {
		super(model, manufacturer);
	}

	public Motherboard(String model, String manufacturer, boolean ecc, String chipset, String formFactor, ArrayList<String> m2slots, int maxMemSupport, int memSlots, ArrayList<String> compatibleMem, String memType, int mSATA_slotCount, ArrayList<String> incEthernetSupp, String incVideo, ArrayList<CountedString> PCISlotList, boolean hasRAID, int sata6gbpsCount, int gen2USBCount, ArrayList<CountedString> gen32USBcount, boolean hasWireless) {
		this(model, manufacturer);
		this.ecc = ecc;
		this.chipset = chipset;
		this.formFactor = formFactor;
		this.m2slots = m2slots;
		this.maxMemSupport = maxMemSupport;
		this.memSlots = memSlots;
		this.compatibleMem = compatibleMem;
		this.memType = memType;
		this.mSATA_slotCount = mSATA_slotCount;
		this.incEthernetSupp = incEthernetSupp;
		this.incVideo = incVideo;
		this.PCISlotList = PCISlotList;
		this.hasRAID = hasRAID;
		this.sata6gbpsCount = sata6gbpsCount;
		this.gen2USBCount = gen2USBCount;
		this.gen32USBcount = gen32USBcount;
		this.isWireless = hasWireless;
	}

	public boolean isEcc() {
		return ecc;
	}

	public void setEcc(boolean ecc) {
		this.ecc = ecc;
	}

	public String getChipset() {
		return chipset;
	}

	public void setChipset(String chipset) {
		this.chipset = chipset;
	}

	public String getFormFactor() {
		return formFactor;
	}

	public void setFormFactor(String formFactor) {
		this.formFactor = formFactor;
	}

	public ArrayList<String> getM2slots() {
		return m2slots;
	}

	public void setM2slots(ArrayList<String> m2slots) {
		this.m2slots = m2slots;
	}

	public int getMaxMemSupport() {
		return maxMemSupport;
	}

	public void setMaxMemSupport(int maxMemSupport) {
		this.maxMemSupport = maxMemSupport;
	}

	public int getMemSlots() {
		return memSlots;
	}

	public void setMemSlots(int memSlots) {
		this.memSlots = memSlots;
	}

	public ArrayList<String> getCompatibleMem() {
		return compatibleMem;
	}

	public void setCompatibleMem(ArrayList<String> compatibleMem) {
		this.compatibleMem = compatibleMem;
	}

	public String getMemType() {
		return memType;
	}

	public void setMemType(String memType) {
		this.memType = memType;
	}

	public int getmSATA_slotCount() {
		return mSATA_slotCount;
	}

	public void setmSATA_slotCount(int mSATA_slotCount) {
		this.mSATA_slotCount = mSATA_slotCount;
	}

	public ArrayList<String> getIncEthernetSupp() {
		return incEthernetSupp;
	}

	public void setIncEthernetSupp(ArrayList<String> incEthernetSupp) {
		this.incEthernetSupp = incEthernetSupp;
	}

	public String getIncVideo() {
		return incVideo;
	}

	public void setIncVideo(String incVideo) {
		this.incVideo = incVideo;
	}

	public ArrayList<CountedString> getPciSlotList() {
		return PCISlotList;
	}

	public void setPciSlotList(ArrayList<CountedString> PCISlotList) {
		this.PCISlotList = PCISlotList;
	}

	public boolean isHasRAID() {
		return hasRAID;
	}

	public void setHasRAID(boolean hasRAID) {
		this.hasRAID = hasRAID;
	}

	public int getSata6gbpsCount() {
		return sata6gbpsCount;
	}

	public void setSata6gbpsCount(int sata6gbpsCount) {
		this.sata6gbpsCount = sata6gbpsCount;
	}

	public int getGen2USBCount() {
		return gen2USBCount;
	}

	public void setGen2USBCount(int gen2USBCount) {
		this.gen2USBCount = gen2USBCount;
	}

	public ArrayList<CountedString> getGen32USBcount() {
		return gen32USBcount;
	}

	public void setGen32USBcount(ArrayList<CountedString> gen32USBcount) {
		this.gen32USBcount = gen32USBcount;
	}

	public boolean isWireless() {
		return isWireless;
	}

	public void setWireless(boolean wireless) {
		isWireless = wireless;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		Motherboard that = (Motherboard) o;

		if (isEcc() != that.isEcc()) return false;
		if (getMaxMemSupport() != that.getMaxMemSupport()) return false;
		if (getMemSlots() != that.getMemSlots()) return false;
		if (getmSATA_slotCount() != that.getmSATA_slotCount()) return false;
		if (isHasRAID() != that.isHasRAID()) return false;
		if (getSata6gbpsCount() != that.getSata6gbpsCount()) return false;
		if (getGen2USBCount() != that.getGen2USBCount()) return false;
		if (isWireless() != that.isWireless()) return false;
		if (getChipset() != null ? !getChipset().equals(that.getChipset()) : that.getChipset() != null)
			return false;
		if (getFormFactor() != null ? !getFormFactor().equals(that.getFormFactor()) : that.getFormFactor() != null)
			return false;
		if (getM2slots() != null ? !getM2slots().equals(that.getM2slots()) : that.getM2slots() != null)
			return false;
		if (getCompatibleMem() != null ? !getCompatibleMem().equals(that.getCompatibleMem()) : that.getCompatibleMem() != null)
			return false;
		if (getMemType() != null ? !getMemType().equals(that.getMemType()) : that.getMemType() != null)
			return false;
		if (getIncEthernetSupp() != null ? !getIncEthernetSupp().equals(that.getIncEthernetSupp()) : that.getIncEthernetSupp() != null)
			return false;
		if (getIncVideo() != null ? !getIncVideo().equals(that.getIncVideo()) : that.getIncVideo() != null)
			return false;
		if (!Objects.equals(PCISlotList, that.PCISlotList))
			return false;
		return getGen32USBcount() != null ? getGen32USBcount().equals(that.getGen32USBcount()) : that.getGen32USBcount() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		result = 31 * result + (isEcc() ? 1 : 0);
		result = 31 * result + (getChipset() != null ? getChipset().hashCode() : 0);
		result = 31 * result + (getFormFactor() != null ? getFormFactor().hashCode() : 0);
		result = 31 * result + (getM2slots() != null ? getM2slots().hashCode() : 0);
		result = 31 * result + getMaxMemSupport();
		result = 31 * result + getMemSlots();
		result = 31 * result + (getCompatibleMem() != null ? getCompatibleMem().hashCode() : 0);
		result = 31 * result + (getMemType() != null ? getMemType().hashCode() : 0);
		result = 31 * result + getmSATA_slotCount();
		result = 31 * result + (getIncEthernetSupp() != null ? getIncEthernetSupp().hashCode() : 0);
		result = 31 * result + (getIncVideo() != null ? getIncVideo().hashCode() : 0);
		result = 31 * result + (PCISlotList != null ? PCISlotList.hashCode() : 0);
		result = 31 * result + (isHasRAID() ? 1 : 0);
		result = 31 * result + getSata6gbpsCount();
		result = 31 * result + getGen2USBCount();
		result = 31 * result + (getGen32USBcount() != null ? getGen32USBcount().hashCode() : 0);
		result = 31 * result + (isWireless() ? 1 : 0);
		return result;
	}

	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeBoolean(ecc);
		dest.writeString(chipset);
		dest.writeString(formFactor);
		dest.writeStringList(m2slots);
		dest.writeInt(maxMemSupport);
		dest.writeInt(memSlots);
		dest.writeStringList(compatibleMem);
		dest.writeString(memType);
		dest.writeInt(mSATA_slotCount);
		dest.writeStringList(incEthernetSupp);
		dest.writeString(incVideo);
		dest.writeParcelableList(PCISlotList, flags);
		dest.writeBoolean(hasRAID);
		dest.writeInt(sata6gbpsCount);
		dest.writeInt(gen2USBCount);
		dest.writeParcelableList(gen32USBcount, flags);
		dest.writeBoolean(isWireless);
	}

	@RequiresApi(Q)
	public Motherboard(@NotNull Parcel in) {
		super(in.readString(), in.readString());
		ecc = in.readBoolean();
		chipset = in.readString();
		formFactor = in.readString();
		m2slots=in.createStringArrayList();
		maxMemSupport = in.readInt();
		memSlots = in.readInt();
		compatibleMem=in.readStringList(compatibleMem);
		memType = in.readString();
		mSATA_slotCount = in.readInt();
		in.readStringList(incEthernetSupp);
		incVideo = in.readString();
		in.readParcelableList(PCISlotList, PCISlotList.getClass().getClassLoader());
		hasRAID = in.readBoolean();
		sata6gbpsCount = in.readInt();
		gen2USBCount = in.readInt();
		in.readParcelableList(gen32USBcount, gen32USBcount.getClass().getClassLoader());
		isWireless = in.readBoolean();
	}

	@NonNull
	@NotNull
	@Override
	public String toString() {
		return "Motherboard " + getModel() + " " + getManufacturer();
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Motherboard.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import org.jetbrains.annotations.Contract
import java.util.*

class Motherboard : Part, Parcelable {
	var isEcc = false // has error correction coding (all chips should tbh)
	var chipset: String? = null
	var formFactor: String? = null
	var m2slots: ArrayList<String>? = null
	var maxMemSupport = 0
	var memSlots = 0
	var compatibleMem: ArrayList<String>? = null
	private var memType: String? = null // pretty much always DDR4
	private var mSATASlotCount = 0 // love SSDs
	var incEthernetSupp: ArrayList<String>? = null
	var incVideo
			: String? = null
	var pciSlotList // had to create a sub-object for this
			: ArrayList<CountedString?>? = null
	var isHasRAID // has "Redundant Array of Inexpensive Disks"
			= false
	var sata6gbpsCount // no clue
			= 0
	var gen2USBCount = 0
	var gen32USBcount // I don't feel like making a new USB object to only be used once
			: ArrayList<CountedString?>? = null
	var wireless: String? = null

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer)
	constructor(
		model: String?,
		manufacturer: String?,
		ecc: Boolean,
		chipset: String?,
		formFactor: String?,
		m2slots: ArrayList<String>?,
		maxMemSupport: Int,
		memSlots: Int,
		compatibleMem: ArrayList<String>?,
		memType: String?,
		mSATA_slotCount: Int,
		incEthernetSupp: ArrayList<String>?,
		incVideo: String?,
		PCISlotList: ArrayList<CountedString?>?,
		hasRAID: Boolean,
		sata6gbpsCount: Int,
		gen2USBCount: Int,
		gen32USBcount: ArrayList<CountedString?>?,
		wireless: String?
	) : this(model, manufacturer) {
		isEcc = ecc
		this.chipset = chipset
		this.formFactor = formFactor
		this.m2slots = m2slots
		this.maxMemSupport = maxMemSupport
		this.memSlots = memSlots
		this.compatibleMem = compatibleMem
		this.memType = memType
		this.mSATASlotCount = mSATA_slotCount
		this.incEthernetSupp = incEthernetSupp
		this.incVideo = incVideo
		pciSlotList = PCISlotList
		isHasRAID = hasRAID
		this.sata6gbpsCount = sata6gbpsCount
		this.gen2USBCount = gen2USBCount
		this.gen32USBcount = gen32USBcount
		this.wireless = wireless
	}

	constructor(`in`: Parcel) : super(`in`.readString(), `in`.readString()) {
		isEcc = `in`.readBoolean()
		chipset = `in`.readString()
		formFactor = `in`.readString()
		m2slots = `in`.createStringArrayList()
		maxMemSupport = `in`.readInt()
		memSlots = `in`.readInt()
		compatibleMem = `in`.createStringArrayList()
		memType = `in`.readString()
		mSATASlotCount = `in`.readInt()
		incEthernetSupp = `in`.createStringArrayList()
		incVideo = `in`.readString()
		pciSlotList = `in`.createTypedArrayList(CountedString.CREATOR)
		isHasRAID = `in`.readBoolean()
		sata6gbpsCount = `in`.readInt()
		gen2USBCount = `in`.readInt()
		gen32USBcount = `in`.createTypedArrayList(CountedString.CREATOR)
		wireless = `in`.readString()
	}

	fun getmSATASlotCount(): Int {
		return mSATASlotCount
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val that = other as Motherboard
		if (isEcc != that.isEcc) return false
		if (maxMemSupport != that.maxMemSupport) return false
		if (memSlots != that.memSlots) return false
		if (getmSATASlotCount() != that.getmSATASlotCount()) return false
		if (isHasRAID != that.isHasRAID) return false
		if (sata6gbpsCount != that.sata6gbpsCount) return false
		if (gen2USBCount != that.gen2USBCount) return false
		if (if (chipset != null) chipset != that.chipset else that.chipset != null) return false
		if (if (formFactor != null) formFactor != that.formFactor else that.formFactor != null) return false
		if (if (m2slots != null) m2slots != that.m2slots else that.m2slots != null) return false
		if (if (compatibleMem != null) compatibleMem != that.compatibleMem else that.compatibleMem != null) return false
		if (if (memType != null) memType != that.memType else that.memType != null) return false
		if (if (incEthernetSupp != null) incEthernetSupp != that.incEthernetSupp else that.incEthernetSupp != null) return false
		if (if (incVideo != null) incVideo != that.incVideo else that.incVideo != null) return false
		if (pciSlotList != that.pciSlotList) return false
		if (if (gen32USBcount != null) gen32USBcount != that.gen32USBcount else that.gen32USBcount != null) return false
		return if (wireless != null) wireless == that.wireless else that.wireless == null
	}

	override fun hashCode(): Int {
		var result = super.hashCode()
		result = 31 * result + if (isEcc) 1 else 0
		result = 31 * result + if (chipset != null) chipset.hashCode() else 0
		result = 31 * result + if (formFactor != null) formFactor.hashCode() else 0
		result = 31 * result + if (m2slots != null) m2slots.hashCode() else 0
		result = 31 * result + maxMemSupport
		result = 31 * result + memSlots
		result = 31 * result + if (compatibleMem != null) compatibleMem.hashCode() else 0
		result = 31 * result + if (memType != null) memType.hashCode() else 0
		result = 31 * result + getmSATASlotCount()
		result = 31 * result + if (incEthernetSupp != null) incEthernetSupp.hashCode() else 0
		result = 31 * result + if (incVideo != null) incVideo.hashCode() else 0
		result = 31 * result + if (pciSlotList != null) pciSlotList.hashCode() else 0
		result = 31 * result + if (isHasRAID) 1 else 0
		result = 31 * result + sata6gbpsCount
		result = 31 * result + gen2USBCount
		result = 31 * result + if (gen32USBcount != null) gen32USBcount.hashCode() else 0
		result = 31 * result + if (wireless != null) wireless.hashCode() else 0
		return result
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		super.writeToParcel(dest, flags)
		dest.writeBoolean(isEcc)
		dest.writeString(chipset)
		dest.writeString(formFactor)
		dest.writeStringList(m2slots)
		dest.writeInt(maxMemSupport)
		dest.writeInt(memSlots)
		dest.writeStringList(compatibleMem)
		dest.writeString(memType)
		dest.writeInt(mSATASlotCount)
		dest.writeStringList(incEthernetSupp)
		dest.writeString(incVideo)
		dest.writeParcelableList(pciSlotList, flags)
		dest.writeBoolean(isHasRAID)
		dest.writeInt(sata6gbpsCount)
		dest.writeInt(gen2USBCount)
		dest.writeParcelableList(gen32USBcount, flags)
		dest.writeString(wireless)
	}

	override val paramCount = 19

	companion object {
		@JvmField
		val CREATOR: Creator<Motherboard> = object : Creator<Motherboard> {
			@Contract("_ -> new")
			override fun createFromParcel(`in`: Parcel) = Motherboard(`in`)

			@Contract(value = "_ -> new", pure = true)
			override fun newArray(size: Int): Array<Motherboard?> = arrayOfNulls(size)
		}
	}
}